### ISSUE TITLE
(Beta) For loop not incrementing

### ISSUE THREAD

**Issue by typicalTYLER** [danielyxie/bitburner#390]
Apologies for the long script, for some reason this script doesn't seem to be operating as I would expect. The main for loop that increments through the servers with the variable i doesn't get incremented after running through the distributedRun calls at the end. It just runs the for loop again with i = 1 forever. However, in v0.39.1 it works correctly without having to manually call i++ at the end. So I think it must be something in the beta.

Another issue that might be related: 
![image](https://user-images.githubusercontent.com/6762150/43275221-dbf39040-90c6-11e8-9745-e2c4320dd979.png)
This script has a warning on the side near the end of the script, but there are no other warnings or errors above.

```
//Skill Multiplier Constants, update after augmentations
hackmult = 1.5566;
growmult = 1.0829;
//Bitnode Multiplier Constants, update after changing Bitnodes
bitnodehackmult = 1.0000;
bitnodegrowmult = 1.0000;
bitnodeweakenmult = 1.0000;

WEAKEN_RAM = 1.75;
GROW_RAM = 1.75;
HACK_RAM = 1.70;

NOT_ROOTED_TOLERANCE = 5;

HOME_SERVER_BUFFER = Math.min(getServerRam("home")[0] / 16, 50); //leave a 16th of the home server ram available on the home server, or 50, whichever is smaller

servers = ["darkweb", "foodnstuff", "sigma-cosmetics", "joesguns", "nectar-net", "hong-fang-tea", "harakiri-sushi", "neo-net", "CSEC", "zer0", "max-hardware", "phantasy", "iron-gym", "silver-helix", "avmnite-02h", "omega-net", "crush-fitness", "johnson-ortho", "the-hub", "comptek", "I.I.I.I", "netlink", "rothman-uni", "aevum-police", "catalyst", "summit-uni", "millenium-fitness", "rho-construction", "run4theh111z", ".", "alpha-ent", "syscore", "lexo-corp", "zb-institute", "snap-fitness", "nova-med", "zb-def", "unitalife", "zeus-med", "titan-labs", "solaris", "global-pharm", "microdyne", "vitalife", "applied-energetics", "deltaone", "omnia", "taiyang-digital", "helios", "galactic-cyber", "univ-energy", "icarus", "omnitek", "aerocorp", "defcomm", "The-Cave", "infocomm", "stormtech", "clarkeinc", "b-and-a", "blade", "powerhouse-fitness", "fulcrumtech", "nwo", "4sigma", "kuai-gong", "ecorp", "megacorp", "fulcrumassets"];

function serverBeingOperatedOn(server) {
    my_servers = ["home"].concat(getPurchasedServers());
    for (i = 0; i < my_servers.length; i++) {
        processes = ps(my_servers[i]);
        for (j = 0; j < processes.length; j++) {
            process = processes[j];
            process_args = args;
            if (process_args.length > 0 && process_args[0] === server) {
                return true;
            }
        }
    }
    return false;
}

function copyCheck(script, target) {
    if (target === "home") {
        // no0op
    } else if (serverExists(target) && !fileExists(script, target)) {
        scp(script, "home", target);
    }
}

function distributedRun(script, threads, arg) {
    print("attempting to distribute " + threads + " threads of " + script + " " + arg);
    remaining_threads = threads;
    while (remaining_threads > 0) {
        my_servers = ["home"].concat(getPurchasedServers());
        for (i = 0; i < my_servers.length; i++) {
            copyCheck(script, my_servers[i]);
            server_ram = getServerRam(my_servers[i]);
            free_ram = server_ram[0] - server_ram[1];
            script_ram = getScriptRam(script, my_servers[i]);
            if (script_ram === 0) {
                print("script_ram is zero for some reason");
                continue;
            }
            if (my_servers[i] === "home") {
                max_threads = Math.floor(Math.max(0, free_ram - HOME_SERVER_BUFFER) / script_ram);
                print("max_threads: " + max_threads);
                server_threads = Math.min(max_threads, remaining_threads);
                if (server_threads <= 0) {
                    continue;
                }
                success = run(script, server_threads, arg);
            } else {
                max_threads = Math.floor(free_ram / script_ram);
                print("max_threads: " + max_threads);
                server_threads = Math.min(max_threads, remaining_threads);
                if (server_threads <= 0) {
                    continue;
                }
                success = exec(script, my_servers[i], server_threads, arg);
            }
            if (success) {
                remaining_threads -= server_threads;
            }
            print("remaining_threads: " + remaining_threads);
            if (remaining_threads <= 0) {
                return true;
            }
        }
        print("finished one loop around my servers, going again");
    }
}

while (true) {
    times_not_rooted = 0;
    for (i = 0; i < servers.length; i++) {
        target = servers[i];
        tprint("server " + i + ": " + target + " out of " + servers.length + " servers");
        //if no root, see if we can get
        if (!hasRootAccess(target)) {
            print("Server: " + target + " does not have root. Continuing...");
            if (target == "darkweb") {
                continue;
            }
            if (times_not_rooted > NOT_ROOTED_TOLERANCE) {
                break;
            } else {
                times_not_rooted++;
            }
        }
        if (getServerMaxMoney(target) === 0) {
            print(target + " can't hold any money.");
            continue;
        }

        //If any scripts running against this server, move along to next server
        if (serverBeingOperatedOn(target)) {
            print(target + " being operated on, continuing...");
            continue;
        }
        //Gather Hack-related Variables
        skill = getHackingLevel();
        reqHack = getServerRequiredHackingLevel(target);
        minsecurity = Math.max(Math.round(getServerBaseSecurityLevel(target) / 3), 1);

        //Calculate number of Hack Threads Required
        perhack = (100 - minsecurity) * ((skill - reqHack + 1) / skill) / 24000 * hackmult * bitnodehackmult;
        hacks = Math.ceil(1 / perhack);

        //Gather Growth-related Variables
        growth = getServerGrowth(target);
        security = minsecurity + hacks * 0.002;
        maxmoney = getServerMaxMoney(target);

        //Calculate number of Grow Threads Required
        growpercent = Math.min(1 + 0.03 / security, 1.0035);
        pergrow = Math.pow(growpercent, growth / 100 * growmult * bitnodegrowmult);
        var1 = maxmoney * Math.log(pergrow);
        lambert = Math.log(var1) - Math.log(Math.log(var1)) - Math.log(1 - Math.log(Math.log(var1)) / Math.log(var1));
        grows = Math.ceil(lambert / Math.log(pergrow));

        //Calculate number of Weaken Threads Required
        weakens = Math.ceil((((hacks * 0.002) + (grows * 0.004)) / (0.05 * bitnodeweakenmult)));
        maxweakens = (100 - minsecurity) / (0.05 * bitnodeweakenmult);
        if (weakens > maxweakens) {
            weakens = maxweakens;
        }

        //Add up how much memory this will use, report the value
        totalmem = Math.round((hacks * HACK_RAM + grows * GROW_RAM + weakens * WEAKEN_RAM + 6.70) * 100) / 100;

        currsecurity = getServerSecurityLevel(target);

        //Weaken to min
        if (currsecurity > minsecurity + 2) {
            initial_weakens = Math.ceil((currsecurity - minsecurity) / 0.05);
            initial_weakens_mem = initial_weakens * WEAKEN_RAM;
            print("Preparing to weaken " + target + " with " + initial_weakens + " weakens for a total of " + initial_weakens_mem + "GB of memory use");
            distributedRun('weaken.script', initial_weakens, target);
        } else {
            print("Preparing to attack " + target + " with " + hacks + " hacks, " + grows + " grows, and " + weakens + " weakens for a total of " + totalmem + "GB of memory use");
            distributedRun('weaken.script', weakens, target);
            distributedRun('grow.script', grows, target);
            distributedRun('hack.script', hacks, target);
        }
        // This i++ is necessary for some reason, because the for loop isn't working.
        i++;
    }
}
```

[forloopsave.zip](https://github.com/danielyxie/bitburner/files/2232794/forloopsave.zip)

**Comment by typicalTYLER** [comment]
In the save the incremented version is called autohackmanager.script and the non incremented version is called autohackmanagers_no_inc.script

**Comment by danielyxie** [comment]
If I had to guess, this is what's happening:

https://stackoverflow.com/questions/2485423/is-using-var-to-declare-variables-optional

**TLDR** Declaring `i = 0` instead of `var i = 0` creates a global variable. So the loops in `distributedRun()` and `serverBeingOperatedOn()` are all changing the same `i` variable, leading to unexpected behavior.

This functionality in beta is actually how it's supposed to work (in JavaScript). Unfortunately this will lead to broken/unexpected behavior in some people's scripts, since the old Netscript didn't do this.

Try adding `var` in front of any variable declarations and let me know if that fixes it.

If this was the problem, thanks for bringing it to my attention. It'll be important to warn people about this change

**Comment by danielyxie** [comment]
Functionality changed in v0.40.0, as discussed above.

Also, as for the "Too many errors" issue, that's related to the script editor's linter. The `Max error count` option in the bottom right of the Script editor of the screen can be increased to avoid this
